blocks/bricks/wrappers.py

```diff
@@ -2,40 +2,91 @@
 from theano import tensor
 
 from blocks.bricks.base import Brick, application
+from blocks.utils import dict_subset, pack
 
 
-class As2D(Brick):
-    """Wraps a 2D application around reshapes to accomodate N-D tensors.
+class WithExtraDims(Brick):
```

Comment:
My main concern is that you trade a simple change to softmax for:
1. introduction of an auxiliary class, which is rather complicated (you 
add 40 lines to do the flattening, even the documentation takes 20 
lines, that is already more than the softmax!)
2. to restore the full ability of axis softmax you need to couple it 
also with the dimshuffling brick -
3. It is very non-intuitive why instead off 1 softmax you suddenly have 
to instantiate it several times, to wrap it different apply methods. 
Then if you instantiate it once, whose children will it be?

In brief: instead of a self-contained, easy to use brick you propose to 
have 2 bricks, one barely simpler (2d softmax), one much more 
complicated (withextradims), and require the end-user to understand the 
intricacies of brick ownership. That doesn't add up for me.

Dmitry Bogdanov wrote:
>
> In blocks/bricks/wrappers.py
> <https://github.com/mila-udem/blocks/pull/759#discussion_r36068241>:
>
> >
> >
> > -class As2D(Brick):
> > -"""Wraps a 2D application around reshapes to accomodate N-D tensors.
> > +class WithExtraDims(Brick):
>
> I guess we are hitting two extremes - the code I pasted has a mask
> argument which is not needed if you allow Softmax to return
> arrays. If you get rid of the masks and just let it operate on a
> chosen axis it will become simpler. In fact in most places you
> replace axis=1 by axis=self.axis. The only non-trivial change is
> the dimshuffle for categorical cross-entropy.
>
> True, but still you have to add reshapes before and after softmax. Why
> not keep the out of the Softmax brick, in another, reusable component?
> Isn't this modularity that we aim for in Blocks?
>
> Similar I don't think that the axis choice should be in Softmax.
> Rather than having all-cases-aware monolith components like your
> |AxisSoftmax| I prefer to have a few components, each with a clear
> simple function (Unix-style, so to say).
>
> The documentation of |Softmax| can indeed be further improved.
>
> Then, blocks should supply the users with good implementations of
> common use-cases. If even we need to handle n-dimensional arrays
> in softmax why not simply add it?
>
> We can add a reference to |WithExtraDims| in softmax documentation.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mila-udem/blocks/pull/759/files#r36068241>.